Publisher/Subscriber (with Broker) implementation

atilag opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

Discussion Issue: #990

### Summary
In order to reduce the coupling of our components, I want to introduce a [Publisher/Subscriber](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern) mechanism in Terra. This is basically an event-driven mechanism where components interact with each other via events/messages.
The basis of this pattern are the following:
* **Publishers** are any component who wants to trigger and event, for example, the transpiler is a publisher who want to notify to all Subscribers his intention to start compiling a circuit, so he sends an event of the type: `terra.transpiler.compile.start`. The transpiler doesn't know anything about the Subscribers, so it's totally decoupled from the rest of the components who want to interact with him.
* **Subscribers** are any component who wants to react to an event. In the example above, the HtmlProgressBar will play the role of a Subscriber, who has subscribed to the event `terra.transpiler.compile.start`, so when this event is triggered, he executes a function (initialize the progress bar).
* The **Broker** is hidden in the Pub/Sub mechanism, it's not supposed to be used by anyone else but Publishers and Subscribers, and it's not exposed. He is the responsible of routing the events published to the subscribers. Is a singleton, so there can only be one instance.


### Details and comments

I have modified the Jupyter progress bar implementation to use the Pub/Sub, as an example of how does it actually work. I think the code self-explanatory but if there's any doubt, yeah, please ask!.
There are two new public classes living in `qiskit/_pubsub.py`:
`class Publisher` and `class Subscriber`.
Being classes means that if we want to turn a component into a Publisher or a Subscriber, the preferred way is by inheriting from these classes:
```python
class MyComponent(Publisher):
    def metod_who_wants_to_publish(self):
      self.publish("my.event")

class MyOtherComponent(Subscriber):
  def __init__(self):
    self.subscribe("my.event", self.my_event_callback)

  def __del__(self):
    self.unsubscribe("my.event", self.my_event_callback)

   def my_event_callback():
        ...

```
If there's no class but we want a method to publish an event, or listen to an event, we can instantiate these classes like:
```python
def method_who_wants_to_publish():
     Publisher().publish("my.event", "some", "arguments")

def method_who_wants_to_listen():
     def callback(some, parameters):
         do_somethin_with_args(some, parameter)

     Subscriber().subscribe("my.event", callback)
```
The Borker implementation is not the canonical implementation, because it doesn't behave like a `message queue`. There's no queue at the moment, because having a queue implies being async, and this something I haven't implemented yet, even though I built part of the scaffolding to transition to this async message queue. See the TODO section below for more information.

### What to decouple
There's no need to decouple every component on Terra, that would be over-engineering, I think that we should be focusing on keeping components with heavy external dependencies like Jupyter tools, or visualization stuff, as decoupled as possible from components that are more susceptible to be reused in all use-cases (like the transpiler, circuits, backends, etc).

### Events hierarchy
Having some sort of event hierarchy for all Terra components can be very beneficial and will keep things readable and simple. My hierarchy proposal:
`"terra.<component>.<method>.<action>"`
Examples:
`"terra.transpiler.compile.start"`
`"terra.job.status.changed"`
`"terra.backend.run.start"`
`"terra.job.result.received"`
...


### Improvements
I think the functionality introduced in this PR is a very good base to start with, but there are things that can be improved, some of them are:

* Make event dispatching asynchronous. The nature of python concurrency model could render this feature useless in most cases, so I'd transition to an async model if and only if there are performance problems derived from Subscribers executing intense computations. 
Otherwise the current blocking system is fine.

* Better unsubscribing mechanism. Having an `unsubscribe_all("event")` could improve code readability in some situations where there are many subscribers to an event.

* Enforce event hierarchy by applying regex to new events. The point is to keep events consistent, allowing them to scale organically, following certain pattern.

* Using decorators for subscribers may also improve user experience:
```python
@subscribe("my.event")
def my_event_callback()
    ...

@subscriber  # <-- Decorating the class so we can automatically unsubscribe
class MyClass:
  @subscribe("my.event")
  def my_event_callback():
    ...

```

### TODO
- [ ] Better in-code documentation
- [ ] Better error handling
- [ ] More tests

atilag commented:
Ready for Review!
@nonhermitian @diego-plan9 @delapuente :)

nonhermitian commented:
Parallel_map should not be publishing as compiler, itis much more general then that. Also, parallel_map is useful to the enduser so parallel module should have underscore removed.

diego-plan9 commented on qiskit/_pubsub.py:
```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2017, IBM.
```
Minor: 2018 :calendar:

diego-plan9 commented on qiskit/_pubsub.py:
```diff
@@ -0,0 +1,126 @@
+
+
+class _Broker(object):
+    """ The event/message broker. It's a singleton."""
```
Also minor - can you  do a round of removing the space after the `"""` in docstrings?

jaygambetta commented:
I need to see a use case of this. Can you make these? As @nonhermitian  points out we need to see this. I still think it would be better to start with the job monitor but I'm fine with this one.

jaygambetta commented:
@mtreinish can you look at this.

nonhermitian commented:
This should not be merged until #1154 is merged, so that we may test the Jupyter parts that are affected by this PR.

atilag commented on qiskit/transpiler/_parallel.py:
```diff
@@ -82,28 +81,16 @@ def parallel_map(task, values, task_args=tuple(), task_kwargs={},  # pylint: dis
         QISKitError: If user interupts via keyboard.
     """
-    # len(values) == 1
+    Publisher().publish("terra.transpiler.compile.start", len(values))
```
I think every component should document the events they emit, so they should be documented in an `Events:` section of the docstring.

atilag commented:
I have addressed all comments here.
@nonhermitian is right, I was introducing transpiler related events into the `parallel_map()` function which is part of the transpiler, but not part of the transpilation in itself. So the way to do that, semantically correct, is by letting `parallel_map()` emmit it's events, so the `transpile_dag()` can subscribe to them, and emmit their own.
The fact that now every component only listen/emmit events, makes composition possible without coupling components, so the re-usability, and flexibility of the entire system has increased a lot.

atilag commented:
Now that #1154 is merged, I'd love to merge this asap.

1ucian0 commented:
This one shoud be first than #975.
